File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/controller/overlay/handler/MarkerHandler.kt
package dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.handler

import com.naver.maps.map.overlay.Marker
import com.naver.maps.map.overlay.Overlay
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asMap
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NMarker
import io.flutter.plugin.common.MethodChannel

internal interface MarkerHandler : OverlayHandler {
    fun handleMarker(
        marker: Overlay, method: String, arg: Any?, result: MethodChannel.Result,
    ) = (marker as Marker).let { m ->
        when (method) {
            NMarker.hasOpenInfoWindowName -> hasOpenInfoWindow(m, result::success)
            NMarker.openInfoWindowName -> arg!!.asMap().let {
                openInfoWindow(m, it["infoWindow"]!!, it["align"]!!, result::success)
            }
            NMarker.positionName -> setPosition(m, arg!!)
            NMarker.iconName -> setIcon(m, arg)
            NMarker.iconTintColorName -> setIconTintColor(m, arg!!)
            NMarker.alphaName -> setAlpha(m, arg!!)
            NMarker.angleName -> setAngle(m, arg!!)
            NMarker.anchorName -> setAnchor(m, arg!!)
            NMarker.sizeName -> setSize(m, arg!!)
            NMarker.captionName -> setCaption(m, arg!!)
            NMarker.subCaptionName -> setSubCaption(m, arg!!)
            NMarker.captionAlignsName -> setCaptionAligns(m, arg!!)
            NMarker.captionOffsetName -> setCaptionOffset(m, arg!!)
            NMarker.isCaptionPerspectiveEnabledName -> setIsCaptionPerspectiveEnabled(m, arg!!)
            NMarker.isIconPerspectiveEnabledName -> setIsIconPerspectiveEnabled(m, arg!!)
            NMarker.isFlatName -> setIsFlat(m, arg!!)
            NMarker.isForceShowCaptionName -> setIsForceShowCaption(m, arg!!)
            NMarker.isForceShowIconName -> setIsForceShowIcon(m, arg!!)
            NMarker.isHideCollidedCaptionsName -> setIsHideCollidedCaptions(m, arg!!)
            NMarker.isHideCollidedMarkersName -> setIsHideCollidedMarkers(m, arg!!)
            NMarker.isHideCollidedSymbolsName -> setIsHideCollidedSymbols(m, arg!!)
            else -> result.notImplemented()
        }
    }

    fun hasOpenInfoWindow(marker: Marker, success: (hasOpenInfoWindow: Boolean) -> Unit)

    fun openInfoWindow(marker: Marker, rawInfoWindow: Any, rawAlign: Any, success: (Any?) -> Unit)

    fun setPosition(marker: Marker, rawPosition: Any)

    fun setIcon(marker: Marker, rawIcon: Any?)

    fun setIconTintColor(marker: Marker, rawIconTintColor: Any)

    fun setAlpha(marker: Marker, rawAlpha: Any)

    fun setAngle(marker: Marker, rawAngle: Any)

    fun setAnchor(marker: Marker, rawNPoint: Any)

    fun setSize(marker: Marker, rawNPoint: Any)

    fun setCaption(marker: Marker, rawCaption: Any)

    fun setSubCaption(marker: Marker, rawSubCaption: Any)

    fun setCaptionAligns(marker: Marker, rawCaptionAligns: Any)

    fun setCaptionOffset(marker: Marker, rawDpOffset: Any)

    fun setIsCaptionPerspectiveEnabled(marker: Marker, rawCaptionPerspectiveEnabled: Any)

    fun setIsIconPerspectiveEnabled(marker: Marker, rawIconPerspectiveEnabled: Any)

    fun setIsFlat(marker: Marker, rawFlat: Any)

    fun setIsForceShowCaption(marker: Marker, rawForceShowCaption: Any)

    fun setIsForceShowIcon(marker: Marker, rawForceShowIcon: Any)

    fun setIsHideCollidedCaptions(marker: Marker, rawHideCollidedCaptions: Any)

    fun setIsHideCollidedMarkers(marker: Marker, rawHideCollidedMarkers: Any)

    fun setIsHideCollidedSymbols(marker: Marker, rawHideCollidedSymbols: Any)
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/controller/overlay/handler/MarkerHandler.swift
import NMapsMap
import Flutter

internal protocol MarkerHandler: OverlayHandler {
    func hasOpenInfoWindow(_ marker: NMFMarker, success: (_ hasOpenInfoWindow: Bool) -> Void)

    func openInfoWindow(_ marker: NMFMarker, rawInfoWindow: Any, rawAlign: Any, success: (Any?) -> Void)

    func setPosition(_ marker: NMFMarker, rawPosition: Any)

    func setIcon(_ marker: NMFMarker, rawIcon: Any?)

    func setIconTintColor(_ marker: NMFMarker, rawIconTintColor: Any)

    func setAlpha(_ marker: NMFMarker, rawAlpha: Any)

    func setAngle(_ marker: NMFMarker, rawAngle: Any)

    func setAnchor(_ marker: NMFMarker, rawNPoint: Any)

    func setSize(_ marker: NMFMarker, rawSize: Any)

    func setCaption(_ marker: NMFMarker, rawCaption: Any)

    func setSubCaption(_ marker: NMFMarker, rawSubCaption: Any)

    func setCaptionAligns(_ marker: NMFMarker, rawCaptionAligns: Any)

    func setCaptionOffset(_ marker: NMFMarker, rawDpOffset: Any)

    func setIsCaptionPerspectiveEnabled(_ marker: NMFMarker, rawCaptionPerspectiveEnabled: Any)

    func setIsIconPerspectiveEnabled(_ marker: NMFMarker, rawIconPerspectiveEnabled: Any)

    func setIsFlat(_ marker: NMFMarker, rawFlat: Any)

    func setIsForceShowCaption(_ marker: NMFMarker, rawForceShowCaption: Any)

    func setIsForceShowIcon(_ marker: NMFMarker, rawForceShowIcon: Any)

    func setIsHideCollidedCaptions(_ marker: NMFMarker, rawHideCollidedCaptions: Any)

    func setIsHideCollidedMarkers(_ marker: NMFMarker, rawHideCollidedMarkers: Any)

    func setIsHideCollidedSymbols(_ marker: NMFMarker, rawHideCollidedSymbols: Any)
}

internal extension MarkerHandler {
    func handleMarker(marker: NMFOverlay,
                      method: String,
                      args: Any?,
                      result: @escaping FlutterResult) {
        let marker = marker as! NMFMarker
        switch method {
        case NMarker.hasOpenInfoWindowName: hasOpenInfoWindow(marker, success: result)
        case NMarker.openInfoWindowName:
            let d = asDict(args!)
            openInfoWindow(marker, rawInfoWindow: d["infoWindow"]!, rawAlign: d["align"]!, success: result)
        case NMarker.positionName: setPosition(marker, rawPosition: args!)
        case NMarker.iconName: setIcon(marker, rawIcon: args)
        case NMarker.iconTintColorName: setIconTintColor(marker, rawIconTintColor: args!)
        case NMarker.alphaName: setAlpha(marker, rawAlpha: args!)
        case NMarker.angleName: setAngle(marker, rawAngle: args!)
        case NMarker.anchorName: setAnchor(marker, rawNPoint: args!)
        case NMarker.sizeName: setSize(marker, rawSize: args!)
        case NMarker.captionName: setCaption(marker, rawCaption: args!)
        case NMarker.subCaptionName: setSubCaption(marker, rawSubCaption: args!)
        case NMarker.captionAlignsName: setCaptionAligns(marker, rawCaptionAligns: args!)
        case NMarker.captionOffsetName: setCaptionOffset(marker, rawDpOffset: args!)
        case NMarker.isCaptionPerspectiveEnabledName: setIsCaptionPerspectiveEnabled(marker, rawCaptionPerspectiveEnabled: args!)
        case NMarker.isIconPerspectiveEnabledName: setIsIconPerspectiveEnabled(marker, rawIconPerspectiveEnabled: args!)
        case NMarker.isFlatName: setIsFlat(marker, rawFlat: args!)
        case NMarker.isForceShowCaptionName: setIsForceShowCaption(marker, rawForceShowCaption: args!)
        case NMarker.isForceShowIconName: setIsForceShowIcon(marker, rawForceShowIcon: args!)
        case NMarker.isHideCollidedCaptionsName: setIsHideCollidedCaptions(marker, rawHideCollidedCaptions: args!)
        case NMarker.isHideCollidedMarkersName: setIsHideCollidedMarkers(marker, rawHideCollidedMarkers: args!)
        case NMarker.isHideCollidedSymbolsName: setIsHideCollidedSymbols(marker, rawHideCollidedSymbols: args!)
        default: result(FlutterMethodNotImplemented)
        }
    }
}
